apps/api/src/app/workflows-v2/util/build-variables.ts

```diff
@@ -41,53 +44,86 @@ export function buildVariables(
   };
 }
 
-/**
- * Validates variables against a schema to identify which ones are valid/invalid.
- * Returns an object containing arrays of valid and invalid variables.
- */
-function identifyUnknownVariables(
+type PathValidationResult = {
+  isValid: boolean;
+  additionalPropertiesFound?: boolean;
+};
+
+function validateVariablePath(variableSchema: Record<string, unknown>, variableName: string): PathValidationResult {
+  const parts = variableName.split('.');
+  let currentPath = 'properties';
+
+  for (const part of parts) {
+    currentPath += `.${part}`;
+    const valueExists = _.get(variableSchema, currentPath) !== undefined;
+    const propertiesPath = `${currentPath}${DOT_PROPERTIES}`;
+    const propertiesExist = _.get(variableSchema, propertiesPath) !== undefined;
+
+    if (!valueExists && !propertiesExist) {
+      const additionalPropertiesResult = checkAdditionalProperties(variableSchema, propertiesPath);
+      if (additionalPropertiesResult.isValid) {
+        return { isValid: true, additionalPropertiesFound: true };
+      }
+
+      return { isValid: false };
+    }
+
+    currentPath = propertiesPath;
+  }
+
+  return { isValid: true };
+}
+
+function checkAdditionalProperties(
   variableSchema: Record<string, unknown>,
-  validVariables: Variable[]
-): TemplateVariables {
-  const validVariablesCopy: Variable[] = _.cloneDeep(validVariables);
+  propertiesPath: string
+): PathValidationResult {
+  let currentPath = propertiesPath;
 
-  const result = validVariablesCopy.reduce<TemplateVariables>(
-    (acc, variable: Variable) => {
-      const parts = variable.name.split('.');
-      let isValid = true;
-      let currentPath = 'properties';
+  while (currentPath.length > 0) {
+    const additionalPropertiesPath = `${currentPath.slice(0, -DOT_PROPERTIES.length)}${DOT_ADDITIONAL_PROPERTIES}`;
+    const additionalPropertiesValue = _.get(variableSchema, additionalPropertiesPath);
 
-      for (const part of parts) {
-        currentPath += `.${part}`;
-        const valueSearch = _.get(variableSchema, currentPath);
+    if (additionalPropertiesValue !== undefined) {
+      return { isValid: additionalPropertiesValue === true };
```

Comment:
the key change, if `additionalPropertiesValue === true` we will allow arbitrary variables.